src/arrays/spinops.jl

```diff
@@ -0,0 +1,124 @@
+#########################
+# Spin Operator Methods #
+#########################
+#########################
+# Implementation  -  1  #
+#########################
+#function spin_h1(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j-1])
+#        N = length(m)+1
+#        m_coeffs = [sqrt(j*(j+1.0) - (x+1.0)*x) for x in m]
+#        return spdiagm(m_coeffs,1,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#function spin_h2(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j])
+#        N = length(m)
+#        return spdiagm(m,0,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#spin_jx(j::Float64) = QuArray(0.5*(spin_h1(j)+spin_h1(j)'))
+#spin_jy(j::Float64) = QuArray(-0.5*im*(spin_h1(j)-spin_h1(j)'))
+#spin_jz(j::Float64) = QuArray(spin_h2(j))
+#spin_jp(j::Float64) = QuArray(spin_h1(j))
+#spin_jm(j::Float64) = QuArray(spin_h1(j)')
+#const sigma_x = 2.0 * spin_jx(0.5)
+#const sigma_y = 2.0 * spin_jy(0.5)
+#const sigma_z = 2.0 * spin_jz(0.5)
+
+# TODO : unicode sigma_y, sigma_z
+#const σₓ = sigma_x
+
+#########################
+# Implementation  -  2  #
+#########################
+
+abstract Spin{S}
+
+immutable HalfSpin{S} <: Spin{S}
+    num::S
+    den::S
+    function HalfSpin(num::S, den::S)
+        if  num>0 && mod(den, 2) == 0 && mod(num, den) == 1
+            new(num, den)
+        else
+            return error("must be half-integer")
+        end
+    end
+end
+HalfSpin{S<:Integer}(n::S, d::S) = HalfSpin{S}(n,d)
+
+immutable IntegerSpin{S} <: Spin{S}
+    num::S
+    function IntegerSpin(num::S)
+        if num>0
+            new(num)
+        else
+            return error("must be integer")
+        end
+    end
+end
+IntegerSpin{S<:Integer}(n::S) = IntegerSpin{S}(n)
+
+function mat_coeffs_1(j::Float64)
+    m = reverse([-j:j-1])
+    N = length(m)+1
+    m_coeffs = [sqrt(j*(j+1.0) - (x+1.0)*x) for x in m]
+    return spdiagm(m_coeffs,1,N,N)
+end
+
+function mat_coeffs_2(j::Float64)
+    m = reverse([-j:j])
+    N = length(m)
+    return spdiagm(m,0,N,N)
+end
+
+function spin_h1(k::HalfSpin)
+    j = float(k.num/k.den)
+    return mat_coeffs_1(j)
+end
+
+function spin_h1(k::IntegerSpin)
+    j = float(k.num)
+    return mat_coeffs_1(j)
+end
+
+function spin_h2(k::HalfSpin)
+    j = float(k.num/k.den)
+    return mat_coeffs_2(j)
+end
+
+function spin_h2(k::IntegerSpin)
+    j = float(k.num)
+    return mat_coeffs_2(j)
+end
+
+spin_jx(j::Union(HalfSpin,IntegerSpin)) = QuArray(0.5*(spin_h1(j)+spin_h1(j)'))
```

Comment:
Instead of the `Union` you can just say `j::Spin`. Since `Spin` is abstract the function will apply to all concrete subtypes; in this case `HalfSpin` and `IntegerSpin`.